### PR TITLE
Extended Resolution Signature

### PR DESCRIPTION
**Pull request by alex-ozdemir**
While extended resolution is a fairly general technique, the paper
"Extended Resolution Simulates DRAT" / the drat2er uses exactly one new
type of rule: definitions of the form

    new <=> old v (~l_1 ^ ~l_2 ^ ... ^ ~l_n)

This PR adds axioms supporting this kind of definition, and adds a test
making use of (most) of those new axioms. The axioms support the following ideas:

   1. Introducing a **fresh** variable, defined in the form above
   2. Clausifying that definition to produce proofs of $$ n + 2 $$ new
      clauses in the form of two clauses, and a cnf with $$ n $$ clauses
   3. Unrolling the proof of the cnf into proofs of the
      original clauses.

You can test this PR by running:
```
lfscc sat.plf er.plf er_test.plf
```

I also had to shuffle around a few other programs to make the dependencies work out.

### CODE REVIEW

File: proofs/signatures/drat.plf [hunk 1]
```diff
@@ -13,17 +13,6 @@
...
-(declare cnfn_proof (cnf_holds cnfn))
-(declare cnfc_proof
-         (! c clause
-         (! deduped_c clause
-            (! rest cnf
-               (! proof_c (holds c)
-                  (! proof_rest (cnf_holds rest)
-                     (! sc (^ (clause_dedup c) deduped_c)
-                        (cnf_holds (cnfc c rest)))))))))
-
```
Comment (alex-ozdemir) [inline proofs/signatures/drat.plf]: This was moved to `sat.plf`.

File: proofs/signatures/lrat.plf [hunk 2]
```diff
@@ -9,12 +9,6 @@
 ;  Some side-conditions make use of a _global assignment_ encoded in
 ;  0 (true) / 1 (false) marks on variables.
 
-; A list of clauses, CNF if interpretted as a formula,
-; but also sometimes just a list
-(declare cnf type)
-(declare cnfn cnf)
-(declare cnfc (! h clause (! t cnf cnf)))
-
```
Comment (alex-ozdemir) [inline proofs/signatures/lrat.plf]: This was moved to `sat.plf`.

File: proofs/signatures/lrat.plf [hunk 3]
```diff
@@ -89,12 +83,6 @@
 ; Functional programs for manipulating types ;
 ; ========================================== ;
 
-; Flip the polarity of the literal
-(program lit_flip ((l lit)) lit
-         (match l
-                ((pos v) (neg v))
-                ((neg v) (pos v))))
-
```
Comment (alex-ozdemir) [inline proofs/signatures/lrat.plf]: This was moved to `sat.plf`.

File: proofs/signatures/lrat.plf [hunk 4]
```diff
@@ -130,28 +118,6 @@
...
-                                   (do (markvar3 v)
-                                     (let result (clc (pos v) (clause_dedup rest))
-                                       (do (markvar3 v) result)))))
-                        ((neg v) (ifmarked4
-                                   v
-                                   (clause_dedup rest)
-                                   (do (markvar4 v)
-                                     (let result (clc (neg v) (clause_dedup rest))
-                                       (do (markvar4 v) result)))))))))
-
```
Comment (alex-ozdemir) [inline proofs/signatures/lrat.plf]: This was moved to `sat.plf`.

File: proofs/signatures/sat.plf [hunk 5]
```diff
@@ -8,10 +8,22 @@
 (declare pos (! x var lit))
 (declare neg (! x var lit))
 
+; Flip the polarity of the literal
+(program lit_flip ((l lit)) lit
+         (match l
+                ((pos v) (neg v))
+                ((neg v) (pos v))))
+
```
Comment (alex-ozdemir) [inline proofs/signatures/sat.plf]: This (and the changes below) were all moved here from other files.

File: proofs/signatures/er.plf [hunk 6]
```diff
@@ -0,0 +1,100 @@
+; Depends on sat.plf
+
+; This file exists to support the **Definition** rule in the paper:
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply proofs/signatures/er.plf]: Right you are! I'll put _definition introduction_ here.

File: proofs/signatures/er.plf [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+;
+;    new <=> old v (~l_1 ^ ~l_2 ^ ... ^ ~l_n)
+;
+; which translates into the clauses:
+;
+;                      new v l_1 v l_2 v ... v l_n
+;                      new v ~old
+;     for each i <= n: l_i v ~new v old
+;
+; This translation could be proven using natural deduction, but instead we make
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply proofs/signatures/er.plf]: Ah, I need to explain further. Making this tool a **theorem**, in the sense I mean it, would not actually slow down proof-checking. I should elaborate on what I mean by **theorem** vs **axiom**.

An axiom is **directly checkable**, **unjustified** statement. By **directly usable**, I mean that during proof-checking, one can check an axiom's use without having to look at anything beyond the statement of the axiom. By **unjustified**, I mean that the reader has to convince themselves that the axiom is true, without any formal justification. Axioms are unjustified at definition-site, and quick to check at use-site.

By a theorem, I mean something that is **directly checkable** and **justified**. That is, a statement that can be quickly (directly) checked by the type checker at use-site, but was proven from other axioms at definition site.

If we could make this tool here a **theorem** in the above sense, then we wouldn't lose any speed, but we would no longer require readers to believe this tool works without justification.

However, I don't really know how to prove/use theorems like this in LFSC, so I made this an axiom instead.

File: proofs/signatures/er.plf [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+; it an axiom for faster checking. It would be cool to make is a theorem, not
+; an axiom!
+
+(declare cnfc_unroll_towards_bottom
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply proofs/signatures/er.plf]: Whoops! That **would** be a good idea. Will do!

File: proofs/signatures/er.plf [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+(program clause_add_suffix_all ((list clause) (suffix clause)) cnf
+         (match list
+                (cln cnfn)
+                ((clc l rest) (cnfc (clc l suffix)
+                                    (clause_add_suffix_all rest suffix)))))
+
+; A definition witness value for:
+;              new <=> old v (~others_1 ^ ~others_2 ^ ... ^ ~others_n)
+; It witnesses the fact that new was fresh when it was defined by the above.
+; Thus it witnesses that the above is TAUT.
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply proofs/signatures/er.plf]: You're seeing the situation clearly -- the mistake is mine. The new definition is not a tautology. It does, however, produce and equisatisfiable formula.

I'll fix the documentation.

File: proofs/signatures/er.plf [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+(declare rat_elimination_def (! new var (! old lit (! others clause type))))
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply proofs/signatures/er.plf]: This is a good question. The idea of extended resolution is that all statements of the form

    fresh <=> function of old variables

can be introduced without modifying satisfiability. Since a RAT-elimination definition has this form, it is justified by the idea of "pure" (would you permit to say _general_?) extended resolution.

However, RAT-elimination definitions are only a subset of the definitions justified by ER. They're a subset that was designed (a) to allow for the elimination of proper RATs and (b) to be easily convertible into clauses. These properties make them useful for converting from DRAT proofs to ER proofs.

So, in conclusion, this kind of definition is related to RATs in that its useful for eliminating them, and is also justified by the general idea of ER.

[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply proofs/signatures/er.plf]: I have a comment at the top of the file that meant to get at this, but I think that comment needs improvement. I'll try to fix it up.

File: proofs/signatures/er_test.plf [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+                    pf_c18
+                  ))))))))))))))))
+                )))
+              )
+            ))
+          )
+       )
+    ))))))))
+  ))))
+)
```
[the remark this answers is not in the document]
Comment (alex-ozdemir) [reply proofs/signatures/er_test.plf]: I came up with a (circuitous) short example.

**Comment by alex-ozdemir** [comment]
Thanks for the review, Yoni!

I believe that I've responded to or address all of your comments. If that isn't the case, let me know. Also let me know if you think that further work should be done to improve any part of this PR!

File: proofs/signatures/er.plf (hunk 10 above)
[the remark this answers is not in the document]
Comment (yoni206) [reply proofs/signatures/er.plf]: This is clearer now, thanks. However, I think that the signature itself will be clearer if it doesn't have lrat/drat notions in it (unless, of course, it is necessary [or, at least, useful] to know that a `definition` originates from a RAT elimination step in the current context).

**Comment by alex-ozdemir** [comment]
Okay, I've modified the names of the various axioms to remove references to the notion of a RAT.

File: proofs/signatures/er.plf [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+; A definition witness value for:
+;              new <=> old v (~others_1 ^ ~others_2 ^ ... ^ ~others_n)
+; It witnesses the fact that new was fresh when it was defined by the above.
+;
+; Thus it witnesses that the above, when added to the formula consisting of the
+; conjunction all the already-proven clauses, produces an equisatisfiable
+; formula.
+(declare definition (! new var (! old lit (! others clause type))))
+
+; Given `old` and `others`, this takes a continuation which expects
```
Comment (yoni206) [inline proofs/signatures/er.plf]: What do you mean by a continuation? Can this be clarified in the comment?

File: proofs/signatures/er.plf [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+;
+;    new <=> old v (~l_1 ^ ~l_2 ^ ... ^ ~l_n)
+;
+; into the clauses:
+;
+;                      new v l_1 v l_2 v ... v l_n
+;                      new v ~old
+;     for each i <= n: l_i v ~new v old              (encoded as (cnf_holds ...))
+(declare clausify_definition
+         (! new var ; elide
```
Comment (yoni206) [inline proofs/signatures/er.plf]: What do these `elide`s explain?

File: proofs/signatures/er_test.plf [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+  (% v2 var
+    (% pf_c1 (holds (clc (pos v1) (clc (pos v2) cln)))
+    (% pf_c2 (holds (clc (neg v1) (clc (pos v2) cln)))
+    (% pf_c3 (holds (clc (pos v1) (clc (neg v2) cln)))
+    (% pf_c4 (holds (clc (neg v1) (clc (neg v2) cln)))
+       (: (holds cln)
+          (decl_definition
+            (neg v1)
+            (clc (neg v2) cln)
+            (\ v3
```
Comment (yoni206) [inline proofs/signatures/er_test.plf]: Consider the following variant of this example:
```
; This is a circuitous proof that uses the definition introduction and
; unrolling rules
(check
  (% v1 var
  (% v2 var
  (% v3 var
    (% pf_dummy (holds (clc (pos v3) (clc (neg v3) cln)))
    (% pf_c1 (holds (clc (pos v1) (clc (pos v2) cln)))
    (% pf_c2 (holds (clc (neg v1) (clc (pos v2) cln)))
    (% pf_c3 (holds (clc (pos v1) (clc (neg v2) cln)))
    (% pf_c4 (holds (clc (neg v1) (clc (neg v2) cln)))
       (: (holds cln)
          (decl_definition
            (neg v1)
            (clc (neg v2) cln)
            (\ v3
            (\ def3
              (clausify_definition _ _ _ def3 _ _
                (\ pf_c5 ; type: (holds (clc (pos v3) (clc (neg v2) cln)))
                (\ pf_c6 ; type: (holds (clc (pos v3) (clc (pos v1) cln)))
                (\ pf_cnf3 ; type: (cnf_holds (cnfc (clc (neg v2) (clc (neg v3) (clc (neg v1) cln))) cnfn))
                  (cnfc_unroll_towards_bottom _ _ pf_cnf3
                    (\ pf_c7 ; type: (clc (neg v2) (clc (neg v3) (clc (neg v1) cln)))
                    (\ pf_cnfn
                                                                          ; Clauses
                  (satlem_simplify _ _ _ (R _ _ pf_c1 pf_c2 v1)  (\ pf_c8 ; v2
                  (satlem_simplify _ _ _ (R _ _ pf_c8 pf_c5 v2)  (\ pf_c9 ; v3
                  (satlem_simplify _ _ _ (R _ _ pf_c9 pf_c7 v3)  (\ pf_c10 ; ~v2  ~v1
                  (satlem_simplify _ _ _ (Q _ _ pf_c10 pf_c8 v2)  (\ pf_c11 ; ~v1
                  (satlem_simplify _ _ _ (Q _ _ pf_c11 pf_c3 v1)  (\ pf_c12 ; ~v2
                  (satlem_simplify _ _ _ (Q _ _ pf_c12 pf_c8 v2)  (\ pf_c13 ; empty
                        pf_c13
                  ))
                  ))
                  ))
                  ))
                  ))
                  ))
                  )))
                )))
              )
            ))
          )
        )
    ))))
  ))))
)
```

The only difference is that I added a dummy `v3` at the beginning. This proof still type-checks, which is good, but I am wondering, can't a similar trick can be used in order to make a bad proof type-check?
Can we (and should we) somehow verify that the fresh variable in the sense of the proof is also fresh in the sense of the lfsc `check` clause?

Comment (alex-ozdemir) [reply proofs/signatures/er_test.plf]: This is a good question. I assume that LFSC has lexical scoping with shadowing, but I don't **know** that.

If it doesn't have proper lexical scoping with shadowing then we should probably fix that. On the other hand, maybe there is some other language mechanism that could be used to implement freshness! I can't think of one right now though.

File: proofs/signatures/er.plf (hunk 13 above)
Comment (yoni206) [reply proofs/signatures/er.plf]: Right now I think it is confusing, so I suggest either to remove them or to clarify in a comment.

**Comment by alex-ozdemir** [comment]
Okay, I'll scrap the elision annotations -- if they're not aiding comprehension then they shouldn't be there.

I'll merge after that.
